lms/djangoapps/grades/new/course_grade_factory.py

```diff
@@ -0,0 +1,186 @@
+        """
+        Returns the CourseGrade for the given user in the course.
+        Reads the value from storage and validates that the grading
+        policy hasn't changed since the grade was last computed.
+        If not in storage, returns a ZeroGrade if ASSUME_ZERO_GRADE_IF_ABSENT.
+        Else, if changed or not in storage, computes and returns a new value.
+
+        At least one of course, collected_block_structure, course_structure,
+        or course_key should be provided.
+        """
+        course_data = CourseData(user, course, collected_block_structure, course_structure, course_key)
+        try:
+            course_grade, read_policy_hash = self._read(user, course_data)
+            if read_policy_hash == course_data.grading_policy_hash:
+                return course_grade
+            read_only = False  # update the persisted grade since the policy changed; TODO(TNL-6786) remove soon
+        except PersistentCourseGrade.DoesNotExist:
+            if assume_zero_if_absent(course_data.course_key):
+                return self._create_zero(user, course_data)
+            read_only = True  # keep the grade un-persisted; TODO(TNL-6786) remove once all grades are backfilled
+
+        return self._update(user, course_data, read_only)
+
+    def read(self, user, course=None, collected_block_structure=None, course_structure=None, course_key=None):
+        """
+        Returns the CourseGrade for the given user in the course as
+        persisted in storage.  Does NOT verify whether the grading
+        policy is still valid since the grade was last computed.
+        If not in storage, returns a ZeroGrade if ASSUME_ZERO_GRADE_IF_ABSENT
+        else returns None.
+
+        At least one of course, collected_block_structure, course_structure,
+        or course_key should be provided.
+        """
+        course_data = CourseData(user, course, collected_block_structure, course_structure, course_key)
+        try:
+            course_grade, _ = self._read(user, course_data)
+            return course_grade
+        except PersistentCourseGrade.DoesNotExist:
+            if assume_zero_if_absent(course_data.course_key):
+                return self._create_zero(user, course_data)
+            else:
+                return None
+
+    def update(self, user, course=None, collected_block_structure=None, course_structure=None, course_key=None):
+        """
+        Computes, updates, and returns the CourseGrade for the given
+        user in the course.
+
+        At least one of course, collected_block_structure, course_structure,
+        or course_key should be provided.
+        """
+        course_data = CourseData(user, course, collected_block_structure, course_structure, course_key)
+        return self._update(user, course_data, read_only=False)
+
+    GradeResult = namedtuple('GradeResult', ['student', 'course_grade', 'err_msg'])
```

Comment:
Nit: This ought to be moved up to the top of the class, I almost missed it tucked away down here